lib/experimental/site-editor-permalinks.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+add_action(
+	'block_editor_settings_all',
+	function ( $settings ) {
+		$settings['__experimentalDashboardLink'] = admin_url();
+		return $settings;
+	}
+);
+
+function gutenberg_remove_query_args( $args ) {
+	$query_string = $_SERVER['QUERY_STRING'];
+	$query        = wp_parse_args( $query_string );
+	foreach ( $args as $arg_name ) {
+		unset( $query[ $arg_name ] );
+	}
+	return $query;
+}
+
+function gutenberg_get_site_editor_url( $path = '', $query = array() ) {
+	$query_string = build_query( array_merge( $query, $path ? array( 'p' => $path ) : array() ) );
+	$base_url     = admin_url( 'site-editor.php' );
+	return $query_string ? $base_url . '?' . $query_string : $base_url;
+}
+
+function gutenberg_get_posts_dataviews_url( $path = '', $query = array() ) {
+	$query_string = build_query(
+		array_merge(
+			$query,
+			$path ? array(
+				'page' => 'gutenberg-posts-dashboard',
+				'p'    => $path,
+			) : array( 'page' => 'gutenberg-posts-dashboard' )
+		)
+	);
+	$base_url     = admin_url( 'admin.php' );
+	return $query_string ? $base_url . '?' . $query_string : $base_url;
+}
+
+function gutenberg_redirect_site_editor_to_design() {
+	global $pagenow;
+	if ( 'site-editor.php' !== $pagenow || isset( $_REQUEST['p'] ) || ! isset( $_SERVER['querystring'] ) ) {
+		return;
+	}
+
+	// The following redirects are for the new permalinks in the site editor.
+	if ( isset( $_REQUEST['postType'] ) && 'wp_navigation' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
+		wp_redirect( gutenberg_get_site_editor_url( '/wp_navigation/' . $_REQUEST['postId'], gutenberg_remove_query_args( array( 'postType', 'postId' ) ) ), 301 );
+		exit;
+	}
```

Comment:
This function would be less repetitive if there were two functions:
- one that returns `false` or a `string` with the new URL
- second one that does the redirect + exit if the result is not `false`

Reply:
Agreed. I'd love to see a more declarative syntax for all those redirects. It could help remove some of the repetition, and would allow for easier maintenance.

Reply:
I don't think a universal function will be prettier, since it will have far too many conditions. 

But some of them can definitely be automated, where this:

```php
if ( isset( $_REQUEST['postType'] ) && 'wp_template' === $_REQUEST['postType'] && ( empty( $_REQUEST['canvas'] ) || empty( $_REQUEST['postId'] ) ) ) {
	wp_redirect( gutenberg_get_site_editor_url( '/template', gutenberg_remove_query_args( array( 'postType' ) ) ), 301 );
	exit;
}

if ( isset( $_REQUEST['postType'] ) && 'wp_template' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
	wp_redirect( gutenberg_get_site_editor_url( '/wp_template/' . $_REQUEST['postId'], gutenberg_remove_query_args( array( 'postType', 'postId' ) ) ), 301 );
	exit;
}

if ( isset( $_REQUEST['postType'] ) && 'wp_block' === $_REQUEST['postType'] && ( empty( $_REQUEST['canvas'] ) || empty( $_REQUEST['postId'] ) ) ) {
	wp_redirect( gutenberg_get_site_editor_url( '/pattern', gutenberg_remove_query_args( array( 'postType' ) ) ), 301 );
	exit;
}

if ( isset( $_REQUEST['postType'] ) && 'wp_block' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
	wp_redirect( gutenberg_get_site_editor_url( '/wp_block/' . $_REQUEST['postId'], gutenberg_remove_query_args( array( 'postType', 'postId' ) ) ), 301 );
	exit;
}

if ( isset( $_REQUEST['postType'] ) && 'wp_template_part' === $_REQUEST['postType'] && ( empty( $_REQUEST['canvas'] ) || empty( $_REQUEST['postId'] ) ) ) {
	wp_redirect( gutenberg_get_site_editor_url( '/pattern', gutenberg_remove_query_args() ), 301 );
	exit;
}

if ( isset( $_REQUEST['postType'] ) && 'wp_template_part' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
	wp_redirect( gutenberg_get_site_editor_url( '/wp_template_part/' . $_REQUEST['postId'], gutenberg_remove_query_args( array( 'postType', 'postId' ) ) ), 301 );
	exit;
}
```

becomes this:

```php
$post_type_redirects = [
	[
		'post_type' => 'wp_template',
		'path' => '/template'
	],
	[
		'post_type' => 'wp_block',
		'path' => '/pattern'
	],
	[
		'post_type' => 'wp_template_part',
		'path' => '/pattern'
	],
];

foreach( $post_type_redirects as $redirect ) {
	if ( isset( $_REQUEST['postType'] ) && $redirect['post_type'] === $_REQUEST['postType'] && ( empty( $_REQUEST['canvas'] ) || empty( $_REQUEST['postId'] ) ) ) {
		wp_redirect( gutenberg_get_site_editor_url( $redirect['path'], gutenberg_remove_query_args() ), 301 );
		exit;
	}

	if ( isset( $_REQUEST['postType'] ) && $redirect['post_type'] === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
		wp_redirect( gutenberg_get_site_editor_url( '/' . $redirect['post_type'] . '/' . $_REQUEST['postId'], gutenberg_remove_query_args( array( 'postType', 'postId' ) ) ), 301 );
		exit;
	}
}
```

Note that this is untested.

Reply:
I can try, all the tests are not similar though, let' see if it works.

Reply:
It's not as simple as the proposal above because of all the special cases and "empty" checks. We can obviously try to make all these checks declarative but we'll be inventing a framework for throwable code. So not worth it IMO. Simpler is better even with duplication.

Reply:
I see and no strong feelings either way. Perhaps some of what Jarda suggests might make it look less repetitive.